Follow up on #1640 - Add ExecutionInfo to RequestTracker callbacks

lukasz-antoniak opened the pull request:
In #1640 we are waiting for author to incorporate our feedback. The intention of this PR is to speed up the process, leaving original author's commit intact and recorded in Git history. Adding `ExecutionInfo` to `RequestTracker` is needed to implement OpenTelemetry support ([JAVA-3148](https://datastax-oss.atlassian.net/browse/JAVA-3148)) and add `onStartXXX()` callbacks ([JAVA-3046](https://datastax-oss.atlassian.net/jira/software/c/projects/JAVA/issues/JAVA-3046)).

Commit:
Add ExecutionInfo to RequestTracker methods

lukasz-antoniak committed:
Add ExecutionInfo to RequestTracker callbacks

lukasz-antoniak commented on core/src/main/java/com/datastax/oss/driver/api/core/tracker/RequestTracker.java:
```diff
@@ -36,38 +37,147 @@
   /**
-   * @deprecated This method only exists for backward compatibility. Override {@link
-   *     #onSuccess(Request, long, DriverExecutionProfile, Node, String)} instead.
+   * Invoked each time a request succeeds.
```
I have reordered methods in the interface, so that most recent versions of all distinct methods are placed first. This shall give a better view on their current state.

tolbertam commented on core/revapi.json:
```diff
@@ -6956,6 +6956,16 @@
         "justification": "Upgrade jackson-databind to 2.13.4.1 to address CVEs, API change cause: https://github.com/FasterXML/jackson-databind/issues/3419"
+      },
+      {
+        "code": "java.method.addedToInterface",
```
I suppose this would not be a backwards compatible change that we are making in a minor release, which [would break the documented following of semver](https://github.com/apache/cassandra-java-driver/blob/a17f7be614a09ab81bc2982b7f7ab3a123b4ab28/manual/api_conventions/README.md?plain=1#L29-L31).

I think it would probably be ok to just make default implementations that return `null`, add a `@Nullable` annotation, and document that callers should consider use of a custom `ExecutionInfo`.  We could mark removing the default implementation for whenever we next do a major release.

I think this would be pragmatic as it would avoid possibly breaking any third party libraries and requiring another release of the library that the user may not have control of.  I'll admit that this seems rather unlikely but I think we should do everything we can to keep semver even if it is somewhat non-optimal.

tolbertam commented on core/src/main/java/com/datastax/oss/driver/internal/core/cql/CqlRequestHandler.java:
```diff
@@ -368,7 +371,8 @@ private void setFinalResult(
       }
     } catch (Throwable error) {
-      setFinalError(callback.statement, error, callback.node, -1);
+      ExecutionInfo executionInfo = defaultExecutionInfo(callback, -1).build();
```
I know that it worked this way before, but it seems odd to me that we would use `-1` for the execution here instead of `callback.execution`.   I guess it's good to leave it this way for consistency like you did, but maybe idea here is "Something unpredictable unexpected happened here that we can't blame on the request itself.

tolbertam reviewed (commented):
Was tentatively +1 on #1640, and I think your changes are also good.  I like that you retained the original commit, made this easier to review and will also allow us to attribute the change to the original contributor.

The only thing I think we should change is just to add default implementations for `getExecutionProfile` on the `ExecutionInfo` interfaces just to retain API compatibility.  With that changed I would be a +1.

tolbertam reviewed (approved):
Looks great, thank you!

lukasz-antoniak committed:
JAVA-3046: Extend RequestTracker interface for observability

Co-authored-by: Andrew Tolbert <6889771+tolbertam@users.noreply.github.com>

lukasz-antoniak committed:
Remove duplicated request tracker callback arguments

lukasz-antoniak commented:
Introduced changes:

1. Add callbacks when we start processing given request (in general and on certain node).
2. Removed duplicated arguments from `RequestTracker` interface. We have been passing the same arguments as method parameters and inside `ExecutionInfo`.

lukasz-antoniak committed:
JAVA-3046: Extend RequestTracker interface for observability
Follow up on #1640 - Add ExecutionInfo to RequestTracker callbacks

Co-authored-by: Andrew Weaver <Andrew.Weaver@wal-mart.com>
Co-authored-by: Andrew Tolbert <6889771+tolbertam@users.noreply.github.com>
Co-authored-by: Jane He <janehesiyao@gmail.com>
